### PR TITLE
Set allowed-origin header based on only valid input requests

### PR DESCRIPTION
**Pull request by kentquirk**
@lornasong think this'll work?

### CODE REVIEW

File: vasco.go [hunk 1]
```diff
@@ -270,16 +270,29 @@ type MatchingReverseProxy struct {
...
+		}
+	}
+
+	if validOrigin {
 		acheaders["Access-Control-Allow-Origin"] = origin.String()
 		acheaders["Vary"] = "Origin"
+	} else {
+		// if the origin wasn't valid, specify something valid
+		// that will cause the browser to fail the request
+		acheaders["Access-Control-Allow-Origin"] = "achievementnetwork.org"
```
Comment (lornasong) [inline vasco.go]: oh! i didn't know this. so, if the request does not originate from achievementnetwork.org but vasco sets that as the origin, the browser will fail it?

File: vasco.go [hunk 2]
```diff
@@ -270,16 +270,29 @@ type MatchingReverseProxy struct {
...
+
 	origin, _ := url.Parse(req.Header.Get("Origin"))
-	log.Printf("URL FOR PROXY: %s %s (%s)", req.Method, req.URL.String(), origin.Host)
+	validOrigin := false
 	if strings.HasPrefix(origin.Host, "localhost") || origin.Host == "" {
+		validOrigin = true
+	}
+	for _, allowed := range f.V.allowedOrigins {
+		if strings.HasSuffix(origin.Host, allowed) {
+			validOrigin = true
```
Comment (lornasong) [inline vasco.go]: should there be a break once set to true? i'm imagining if an ".achievementnetwork.org" client makes a request, and the last checked allowedOrigin is ".anettech.org", will validOrigin become false?

Comment (kentquirk) [inline vasco.go]: no, there's never a way to set to false again -- but a break wouldn't hurt to avoid redundant tests.

Comment (lornasong) [inline vasco.go]: bah :facepalm: ok, i see what you're saying and where i'm getting it wrong. thanks for clarifying!

File: vasco.go (hunk 1 above)
Comment (kentquirk) [inline vasco.go]: That's the point of allow-origin, yeah. Basically, on each request, the browser asks "hey, I got code from www.foo.com, is that OK for that code to talk to you?". We could respond either with a 403 ("forbidden") or with 200 and "here's an origin that's valid for us". The browser will notice that it's not the one that was sent.

Comment (lornasong) [inline vasco.go]: thanks! i am learning a lot about this origin and browser stuff (i did not realize the browser did so much).

**Pull request merged**